### ISSUE TITLE
where is the gradient computation of a variable in cpu

### ISSUE THREAD

**Issue by iamhankai** [tensorflow/tensorflow#7772]
There are some variables of a neural network. Some are in GPU, and some are in CPU. For the variables in CPU, where are their gradients computed? CPU or GPU?
For example, where is the gradient of `c` computed?
```python
with tf.device('/gpu:0'):
  a = tf.get_variable("a", [100, 3])
  b = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[3, 2])
  with tf.device('/cpu:0'):
    c = tf.matmul(a, b)
  d = tf.matmul(b,c)
```
Thanks very much.

**Comment by cjliux** [comment]
Eh.....CPU, I guess, because tools like this almost always prefer being lazy in construction and aggressive in evaluation.

**Comment by cjliux** [comment]
An interesting problem...

**Comment by shamoya** [comment]
You can see all the variables and operations placement by adding config=tf.ConfigProto(log_device_placement=True) to session creation.
The gradient computation operations will probably be placed on the device where the optimzer was placed (you don't have it in your example).
BTW your example doesn't make sense, b and c are compatible for matmul, and also the device placement are nested.

**Comment by sskgit** [comment]
From the documentation at https://www.tensorflow.org/tutorials/using_gpu

If a TensorFlow operation has both CPU and GPU implementations, the GPU devices will be given priority when the operation is assigned to a device. For example, matmul has both CPU and GPU kernels. On a system with devices cpu:0 and gpu:0, gpu:0 will be selected to run matmul.

# Creates a graph.
a = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[2, 3], name='a')
b = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[3, 2], name='b')
c = tf.matmul(a, b)
# Creates a session with log_device_placement set to True.
sess = tf.Session(config=tf.ConfigProto(log_device_placement=True))
# Runs the op.
print sess.run(c)

**Comment by iamhankai** [comment]
I have used  config=tf.ConfigProto(log_device_placement=True) to see the device placement log. The gradients are computed in GPU preferentially.
